File: app/src/main/java/com/semisonfire/cloudgallery/ui/photo/PhotoDetailPresenter.java
package com.semisonfire.cloudgallery.ui.photo;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.semisonfire.cloudgallery.core.presentation.BasePresenter;
import com.semisonfire.cloudgallery.data.model.Photo;
import com.semisonfire.cloudgallery.data.remote.RemoteRepository;
import com.semisonfire.cloudgallery.di.ActivityScope;
import com.semisonfire.cloudgallery.utils.FileUtils;

import java.net.URL;

import javax.inject.Inject;

import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@ActivityScope
public class PhotoDetailPresenter extends BasePresenter<PhotoDetailContract.View>
        implements PhotoDetailContract.Presenter {

    private RemoteRepository remoteRepository;

    @Inject
    public PhotoDetailPresenter(RemoteRepository remoteRepository) {
        this.remoteRepository = remoteRepository;
    }

    @Override
    public void download(Photo photo) {
        getCompositeDisposable().add(
                remoteRepository.getDownloadLink(photo)
                        .map(link -> {
                            URL url = new URL(link.getHref());
                            Bitmap bitmap = BitmapFactory
                                    .decodeStream(url.openConnection().getInputStream());
                            return FileUtils.getInstance().savePublicFile(bitmap, photo.getName());
                        })
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                path -> getView().onPhotoDownloaded(path),
                                throwable -> getView().onError(throwable)
                        ));
    }

    @Override
    public void delete(Photo photo, int from) {

        Flowable<Photo> delete;

        switch (from) {
            case PhotoDetailActivity.FROM_DISK:
                delete = remoteRepository.deletePhoto(photo);
                break;
            case PhotoDetailActivity.FROM_TRASH:
                delete = remoteRepository.deleteTrashPhoto(photo);
                break;
            default:
                delete = Flowable.just(photo);
        }

        getCompositeDisposable().add(
                delete.subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                p -> getView().onFilesChanged(p),
                                throwable -> getView().onError(throwable)
                        ));
    }

    @Override
    public void restore(Photo photo) {
        getCompositeDisposable().add(
                remoteRepository.restoreTrashPhoto(photo)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                p -> getView().onFilesChanged(p),
                                throwable -> getView().onError(throwable)
                        ));
    }

    @Override
    public void createShareFile(Bitmap bitmap) {
        getCompositeDisposable().add(
                Single.just(bitmap)
                        .map(btmp -> FileUtils.getInstance().createShareFile(btmp))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                uri -> getView().onFilePrepared(uri),
                                throwable -> getView().onError(throwable)
                        ));
    }
}

File: app/src/main/java/com/semisonfire/cloudgallery/ui/detail/PhotoDetailPresenter.kt
package com.semisonfire.cloudgallery.ui.detail

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.semisonfire.cloudgallery.core.ui.Presenter
import com.semisonfire.cloudgallery.data.model.Photo
import com.semisonfire.cloudgallery.logger.printThrowable
import com.semisonfire.cloudgallery.ui.disk.data.DiskRepository
import com.semisonfire.cloudgallery.utils.FileUtils
import com.semisonfire.cloudgallery.utils.background
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.PublishSubject
import java.net.URL
import javax.inject.Inject

interface PhotoDetailPresenter : Presenter {

    fun download(photo: Photo)
    fun createShareFile(bitmap: Bitmap)

    fun observePhotoDownloaded(): Observable<String>
    fun observeFilePrepared(): Observable<Uri>
}

class PhotoDetailPresenterImpl @Inject constructor(
    private val diskRepository: DiskRepository
) : PhotoDetailPresenter {

    private val compositeDisposable = CompositeDisposable()

    private val photoDownloadListener = PublishSubject.create<String>()
    private val filePreparedListener = PublishSubject.create<Uri>()

    override fun observePhotoDownloaded(): Observable<String> {
        return photoDownloadListener
    }

    override fun observeFilePrepared(): Observable<Uri> {
        return filePreparedListener
    }

    override fun download(photo: Photo) {
        compositeDisposable.add(
            diskRepository
                .getDownloadLink(photo.remotePath)
                .map {
                    val url = URL(it.href)
                    val bitmap = BitmapFactory.decodeStream(url.openConnection().getInputStream())
                    FileUtils.savePublicFile(bitmap, photo.name)
                }
                .subscribeOn(background())
                .subscribe(
                    { photoDownloadListener.onNext(it) },
                    { it.printThrowable() }
                )
        )
    }

    override fun createShareFile(bitmap: Bitmap) {
        compositeDisposable.add(
            Single.just(bitmap)
                .map { FileUtils.createShareFile(it) }
                .subscribeOn(background())
                .subscribe(
                    { filePreparedListener.onNext(it) },
                    { it.printThrowable() }
                )
        )
    }

    override fun dispose() {
        compositeDisposable.clear()
    }
}